### PR TITLE
package 3.0

### PR DESCRIPTION
**Pull request by remstone7**
new version:

1. Refactor all the code and establish a new pattern
2. Add unit tests
3. Update README.md

### CODE REVIEW

File: klaviyo/api_helper.py [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+            return response.text
+
+    def __handle_200_response(self, response, status_code):
+        """Determines how to handle a 200 http response.
+
+        Args:
+            response (obj): Requests object.
+            status_code (int): HTTP status code.
+        Returns:
+            (json or http ok)
```
Comment (smoucka) [inline klaviyo/api_helper.py:249]: can update that this returns a KlaviyoAPIResponse object

File: klaviyo/api_helper.py [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+
+        Args:
+            params (dict): API Query params.
+
+        Returns:
+            (dict): Valid query parameters.
+        """
+        return dict((k, v) for k, v in params.items() if v is not None)
+
+    def _build_marker_param(self, marker):
```
Comment (klaviyojad) [inline klaviyo/api_helper.py:103]: missing docsring to explain what marker is and what this method returns

File: klaviyo/api_helper.py [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+            (dict or arr): Response from Klaviyo API.
+        """
+
+        url = '{}/{}/{}'.format(
+            self.api_server,
+            self.V2_API,
+            path,
+        )
+        data.update({
+            'api_key': self.private_token
```
Comment (smoucka) [inline klaviyo/api_helper.py:160]: can constant 'api_key'

File: klaviyo/api_helper.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+            params (dict): Query parameters for the api call.
+        Returns:
+            (dict or arr): Response from Klaviyo API.
+        """
+        url = '{}/{}/{}'.format(
+            self.api_server,
+            self.V1_API,
+            path,
+        )
+        formatted_url = '{}?api_key={}'.format(url, self.private_token)
```
Comment (smoucka) [inline klaviyo/api_helper.py:181]: update params instead of formatting string

File: klaviyo/exceptions.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
+import simplejson
+
+
+class KlaviyoException(Exception):
+    def __init__(self, message):
+        super(KlaviyoException, self).__init__(message)
+        self.message = message
+
+
+class KlaviyoConfigurationException(Exception):
```
Comment (smoucka) [inline klaviyo/exceptions.py:10]: parent shoudl be klaviyoexception

File: klaviyo/exceptions.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        Returns:
+            A dict or text of the response failure.
+        """
+        message = str()
+        try:
+            message = response.json()
+        except (simplejson.JSONDecodeError, ValueError) as e:
+            message = response.text
+        return message
+
```
Comment (smoucka) [inline klaviyo/exceptions.py:36]: remove third newline

File: klaviyo/lists.py [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    LIST = 'list'
+    LISTS = 'lists'
+    SUBSCRIBE = 'subscribe'
+    MEMBERS = 'members'
+    ALL = 'all'
+    LIST_NAME = 'list_name'
+
+    def get_lists(self):
+        """Returns a list of Klaviyo lists
+        https://www.klaviyo.com/docs/api/v2/lists#post-lists
```
Comment (smoucka) [inline klaviyo/lists.py:14]: double nit: incorrect url and a newline after summary docstring between additional info

File: klaviyo/lists.py [hunk 8]
```diff
@@ -0,0 +1,214 @@
...
+    def add_subscribers_to_list(self, list_id, profiles):
+        """Uses the subscribe endpoint to subscribe user to list, this obeys the list settings.
+        https://www.klaviyo.com/docs/api/v2/lists#post-subscribe
+
+        Args:
+            list_id (str): The list id.
+            profiles (list of dict): List of dicts containing profile info.
+                [{
+                    'email': email_address,
+                    // key:value profile properties
```
Comment (smoucka) [inline klaviyo/lists.py:87]: this can be a python comment not a js  comment

File: klaviyo/lists.py [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+    def add_members_to_list(self, list_id, profiles):
+        """Adds a member to a list regardless of opt-in settings.
+        https://www.klaviyo.com/docs/api/v2/lists#post-members
+
+        Args:
+            list_id (str): The list id.
+            profiles (list of dict): List of dicts containing profile info.
+                [{
+                    'email': email_address,
+                    // key:value profile properties
```
Comment (smoucka) [inline klaviyo/lists.py:141]: #

File: klaviyo/lists.py [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+        Returns:
+            (list) of dicts corresponding to the email addresses on their list if they're on the list
+        """
+        params = {
+            self.EMAILS: emails
+        }
+
+        return self._v2_request('{}/{}/{}'.format(self.LIST, list_id, self.MEMBERS), self.HTTP_GET, params)
+
+    def delete_list_membership_status(self, list_id, emails):
```
Comment (smoucka) [inline klaviyo/lists.py:169]: update method name to 'remove_members_from_list'

File: klaviyo/lists.py [hunk 11]
```diff
@@ -0,0 +1,214 @@
...
+        }
+
+        return self._v2_request('{}/{}/{}'.format(self.LIST, list_id, self.MEMBERS), self.HTTP_DELETE, params)
+
+    def get_list_exclusions(self, list_id, marker=None):
+        """Get all of the emails that have been excluded from a list along with the exclusion reason and exclusion time.
+        https://www.klaviyo.com/docs/api/v2/lists#get-exclusions-all
+
+        Args:
+            list_id (str):The list id.
```
Comment (smoucka) [inline klaviyo/lists.py:191]: space between colon and description

File: klaviyo/metrics.py [hunk 12]
```diff
@@ -0,0 +1,121 @@
+from .api_helper import KlaviyoAPI
+
+STARTING_PAGE = 0
+DEFAULT_BATCH_SIZE = 50
+TIMELINE_BATCH_SIZE = DEFAULT_BATCH_SIZE + DEFAULT_BATCH_SIZE
```
Comment (smoucka) [inline klaviyo/metrics.py:5]: move into class

File: klaviyo/metrics.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+            self.MEASUREMENT: measurement,
+            self.WHERE: where,
+            self.BY: by,
+            self.COUNT: count
+        }
+        params = self._filter_params(params)
+
+        url = '{}/{}/{}'.format(self.METRIC, metric_id, self.EXPORT)
+
+        return self._v1_request(url, self.HTTP_GET, params)
```
Comment (smoucka) [inline klaviyo/metrics.py:121]: newline at end of file

File: klaviyo/public.py [hunk 14]
```diff
@@ -0,0 +1,129 @@
...
+
+        Args:
+            email (str or None): Email address.
+            id (str or None): External id for customer.
+            properties (dict): Information about the customer.
+            is_test (bool): Should this be a test request.
+        Returns:
+            (str): 1 (pass) or 0 (fail)
+        """
+        if email is None and id is None:
```
Comment (smoucka) [inline klaviyo/public.py:115]: if not email and not id

File: tests/fixtures/api_helper.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+
+    @pytest.fixture
+    def mock_requests_package(self):
+        with patch.object(requests, requests.get.__name__) as mock_requests_package:
+            yield mock_requests_package
+
+
+    @pytest.fixture
+    def mock_url(self):
+        return ' https://a.klaviyo.com/api/v2/lists'
```
Comment (smoucka) [inline tests/fixtures/api_helper.py:79]: remove space

File: .gitignore [hunk 16]
```diff
@@ -20,8 +20,15 @@ pip-log.txt
...
+
 #Translations
 *.mo
 
 #Mr Developer
-.mr.developer.cfg
\ No newline at end of file
+.mr.developer.cfg
+
+.DS_Store
```
Comment (jordanallain) [inline .gitignore:34]: new line?

File: klaviyo/exceptions.py [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+    def __init__(self, message):
+        super(KlaviyoException, self).__init__(message)
+        self.message = message
+
+
+class KlaviyoConfigurationException(KlaviyoException):
+    pass
+
+
+class KlaviyoApiException(Exception):
```
Comment (jordanallain) [inline klaviyo/exceptions.py:14]: Probably a nit but should this be `KlaviyoAPIException`?

API vs Api

the other classes seem to use API (KlaviyoAPI, KlaviyoAPIFixture), not a big deal but consistency is good.

Comment (remstone7) [reply klaviyo/exceptions.py:14]: totally 👍

**Pull request merged**